### DISCUSSION TITLE
System tags

### DISCUSSION THREAD

**Discussion by majkinetor** [1Remote/1Remote#416] [Ideas]
The idea is to use temporary system tags that could be added next to user tags but are maintained by the system automatically.  We could have them as a mechanism for different stuff:

1. **Running connections**
Currently running connections should be marked as running. Then you can select them all in the list quickly, and you will also see a status when launching multiple connections via main window. This is sorta like 'currently open windows' in an editor.
3. **Last disconnection status**
Show them when database is disconnected or when latest disocverability test fails or when user runs bulk discoverability test (it does not exist now but should IMO). 1RM shouldn't IMO check this ocassionally, but it would be lost on restart or succesifull test.

Those tags could otherwise behave the same as normal tags. You would for get imediatelly an option to 1) export running connections 2) edit them (add tags etc) and 3) delete them, something we don't have now. You will get an option to pin running connections and disconnections etc. More possibilities would come in the future, for sure. Maybe even some stuff could be refactored to use system tags, such as protocol (instead having protocol column which is next to usless now, we cold have protocol tag, which is very useful).

<details>
Here how latest disconnection status could look like:

![image](https://github.com/1Remote/1Remote/assets/85767/78e605bf-29c3-4934-b334-d180fa4d5aea)
</details>

**Comment by majkinetor** [comment]
I have this on online banking system I lead and people really like it:

<details><summary>details</summary> 
System marks executed payment orders and file id it comes from. User can put his own tags. Tags are used in search and can be clicked as usual.

![image](https://github.com/1Remote/1Remote/assets/85767/38207e43-36fc-4acd-b446-9058aa182a58)
</details>

**Reply by VShawn** [reply]
I would like to know what system labels there will be. 

Annnnd, there is not enough space for many tags to be displayed before we refactor the list interface, so I suggest that we redesign the list page before addressing this requirement.

**Comment by majkinetor** [comment]
It makes sence to have:

1. Active/running 
2. Protocol (just one at a time): rdp | ssh | telnet | ftp |  sftp | app | remote-app (7 total + maybe more unique, user addable for custom protocols, for example free-rdp)
3. Latest availability status.
4. Database tag makes sense too, not sure, maybe configurable (show databases as tags vs interface groups)

So far 3 or 4 tags in total. But good UI should be designed so it allows adding more, since such feature opens up possibility to set any kind of status/meta info there, as it gives you quick way to show only connections having specific thing on it. For example, imagin we add username as a tag - you imediatelly got new feautre - show only connections having specific username, export, bulk connect, delete.  Sure, nothing stops you to do that now with user tags, but some of those would be hard to maintain on dynamic things, and some are impossible (runnig/availability) plus some will be wrong (change database or connection protocol and you must change their tag to match).

[MOVED UI DISCSSION TO #419]

System tags could be colored differently (maybe optional, on picture I used only red color, but you could have lets say light blue for the protocol, red for disconnection etc.) so you can more easily identify certain tags.

**Reply by VShawn** [reply]
1、2、3 is ok

but 4 

> Database tag makes sense too, not sure, maybe configurable (show databases as tags vs interface groups)

I don't like `database tag`, group is easier for me to find a connection. 

And make it configurable will take a lot of work on UI.

**Reply by VShawn** [reply]
> Here is one possible mockup I created now

It seems that each connection has a different height, which appears to affect the virtualization of the WPF listview. I'm not sure, as I've mentioned I'm not a UI expert.

**Reply by majkinetor** [reply]
Having a db as a tag gives better sorting options. It also simplifies the interface as without it you don't have groups but flat list. Its not that important IMO.

> It seems that each connection has a different height, which appears to affect the virtualization of the WPF listview.

OK, so maybe we ditch WPF and use HTML instead? In HTML all this is trivial. 

I will try to find time to see what can be done, I need to setup dev environment first. But since you are OK with this I will make ticket for other stuff (put👍 on the first post please, that will be the main marker, 👍 on particular note means you are ok with that note, and on the first post that you think we should implement this idea as it was resolved in the discussion).

**Reply by majkinetor** [reply]
When I say go with this, you mean both system tags and new UI ? Or just system tags with current UI. Currently there are not so many system tags that we can't add them in the current UI. We can then also remove Protocol column and get some small space.

**Reply by VShawn** [reply]
> Having a db as a tag gives better sorting options

I don't think so, the sorting options is not that good, searching is best way to find a connection.

When I'm just browsing through them, grouping connections together is the most comfortable way to me.

Putting connections from different databases together makes me feel confused, even though I know I can sort them by their db tags.

We can make it grouped when no searching keyword, no group when searching #413 

> 👍 on particular note means you are ok with that note, 

You mean only this one?

![image](https://github.com/1Remote/1Remote/assets/10143738/f211cee8-29fc-463f-8ab8-a50b49802298)

Or each one?

![image](https://github.com/1Remote/1Remote/assets/10143738/62cd4cdb-b0da-42dc-96b1-b5e03ea015e4)



What should I do if I only agree with half of it?

And what should I do if I not ok with the following discussion?

> use HTML instead? In HTML all this is trivial.

Electron has poor performance, I don't like it. a small feature takes 100MB of space and even more memory.


> you mean both system tags and new UI

I mean `just system tags with current UI`....working on the UI takes too much time, and there is no sense of accomplishment:(

We can move forward with this discussion step by step, such as removing the protocol column, adding a protocol tag and a running tag first, and then considering the hardcore part(new ui and whether db tag or not) when everything is ready.

**Reply by majkinetor** [reply]
> I don't think so, the sorting options is not that good, searching is best way to find a connection.

Don't be ridiculous, you can't impose your _own_ workflow on the World. People may need to sort for various reasons, reporting and double checking is one of them, not only to find something.

> When I'm just browsing through them, grouping connections together is the most comfortable way to me.

As we are doing enterprise app now, you need to stop thinking in those terms :) Its not personal any more 

> Putting connections from different databases together makes me feel confused, even though I know I can sort them by their db tags.

Sure, that is your case. Case of others may be different. Maybe they will split databases by project and want to see all the "web servers" across all the projects as an example. Or split them by region. Sorted :)

> We can make it grouped when no searching keyword, no group when searching https://github.com/1Remote/1Remote/discussions/413

Sure we can, but then people like you will get confused, lol :) I want to support all use cases, yours private small use included :)

But like I said, perhaps benefit is not big enough. It can be done later too if people want it. 

> You mean only this one?

Yeah, the [first one](https://github.com/1Remote/1Remote/discussions/416#discussion-5193030). 

> Or each one?

That is up to you if you want, you can mark _parts of the discussion you think are OK_ and thumb down them if you don't think, but that is just informative. At the end, I would need a marker of final status on the first note so I can know if the entire discussion was agreed upon or not. This is a voting system, we currently have 2 votes available for all ideas, in the future we might have more :) I will document this as a part of CONTRIBUTE.md #406 .

> And what should I do if I not ok with the following discussion?

You say it, or even 👎 , but more importantly say your opinions. 

> Electron has poor performance, I don't like it. a small feature takes 100MB of space and even more memory.

For sure. I wasn't thinking about electron at all, more something like [Ultralight](https://ultralig.ht) - only 10mb. Not saying it should be that one, or [RmlUI](https://github.com/mikke89/RmlUi) or [Sciter](https://sciter.com/) (5MB) but this is just on top of my head, I didn't research this recently.

> I mean just system tags with current UI....working on the UI takes too much time, and there is no sense of accomplishment:(

OK, I agree with you on this, particularly given your stance on UI :)

If I mange to make something on my own it I will publish it in specific branch.

**Reply by majkinetor** [reply]
UI discussion now moved to #419

**Reply by VShawn** [reply]
> Don't be ridiculous, you can't impose your own workflow on the World

I agree with this point, so I won't remove db grouping without our user demand research.

like you may not know that 25% of PRM users were using the card view every day before research.
